yo1k/tic_tac_toe/core/game.py

```diff
@@ -0,0 +1,296 @@
+
+    def _clean(self):
+        self.__cells = Board.__empty_cells()
+
+    @property
+    def cells(self):
+        return self.__cells
+
+    @staticmethod
+    def size() -> int:
+        return 3
+
+    @staticmethod
+    def __empty_cells() -> list[list[Optional[Mark]]]:
+        return [[None for _ in range(Board.size())] for _ in range(Board.size())]
+
+    def __str__(self):
+        return f"{type(self)}(" \
+               f"cells={self.__cells},"
+
+
+class State:
+    """Full game state which is enough to restore a paused game."""
+
+    def __init__(self,
+                 game_rounds: int,
+                 players: Sequence[Player],
+                 board: Board,
+                 phase=Phase.ONSET,
+                 cur_round: int = 0,
+                 step: int = 0):
+        self.game_rounds = game_rounds
+        self.round = cur_round
+        self.players = players
+        self.step = step
+        self.phase = phase
+        self.board = board
+
+    def __str__(self):
+        return f"{type(self)}(" \
+               f"game_rounds={self.game_rounds}," \
+               f"round={self.round}," \
+               f"players={self.players}," \
+               f"step={self.step}," \
+               f"phase={self.phase}," \
+               f"board={self.board})"
+
+
+class Action:
+    def __init__(self, surrender: bool, occupy: Optional[Cell], next_round: bool):
+        assert (surrender is True and occupy is None and next_round is False) or \
+               (surrender is False and occupy is not None and next_round is False) or \
+               (surrender is False and occupy is None and next_round is True)
+        self.__surrender = surrender
+        self.__occupy = occupy
+        self.__start = next_round
+
+    @staticmethod
+    def new_surrender() -> "Action":
+        return Action(True, None, False)
+
+    @staticmethod
+    def new_occupy(occupy: Cell) -> "Action":
+        return Action(False, occupy, False)
+
+    @staticmethod
+    def new_start():
+        return Action(False, None, True)
+
+    @property
+    def surrender(self):
+        return self.__surrender
+
+    @property
+    def occupy(self):
+        return self.__occupy
+
+    @property
+    def start(self):
+        return self.__start
+
+    def __str__(self):
+        if self.__surrender is True:
+            action = "surrender"
+        elif self.__occupy is not None:
+            action = f"new_occupy={self.__occupy}"
+        elif self.__start is True:
+            action = "start"
+        else:
+            assert False
+        return f"{type(self)}(" \
+               f"{action})"
+
+
+class ActionQueue(ABC):
+    @abstractmethod
+    def next(self) -> Optional[Action]:
+        pass
+
+
+class Logic:
+    """Game logic."""
+
+    def __init__(self, action_queues: Sequence[ActionQueue]):
+        """Indexes in `action_queues` correspond to indexes in `State.players`."""
+        self.__action_queues = action_queues
```

Comment:
Let's add an assertion that the size of `action_queues` is 2.

Reply:
Done.

Reply:
I realized that since we have two assertions (for action queues and for players) that use the same expectation, we should make 2 a constant by introducing a static method `State.playersCount`, and then use it in both assertions.

Reply:
Done.